meet.py should consider all Tuple[t1, t2, ...] as subtypes of plain tuple

ilevkivskyi opened the pull request:
Fixes #1691 
Fixes #2456 

This is a simple-minded fix. In the future we could add more rules for tuples in ``meet.py`` (for example for homogeneous tuples, a tuple and an iterable etc).

Tests are based on (extended) examples discussed in issues.

JukkaL commented:
Thanks for the fix!

I'm going to postpone merging this until after 0.4.6 as we are very close to the release.

ilevkivskyi commented:
@JukkaL Just pinging this. (I am working on other thing that would be easier with this fix merged.)

JukkaL commented on mypy/meet.py:
```diff
@@ -243,6 +245,10 @@ def visit_tuple_type(self, t: TupleType) -> Type:
             # TODO: What if the fallbacks are different?
             return TupleType(items, t.fallback)
+        # Special case: all Tuple[t1, t2, ...] are subtypes of plain tuple.
+        elif (isinstance(self.s, Instance) and self.s.type.fullname() == 'builtins.tuple'
```
We could also handle cases like the meet of `Tuple[int, int]` and `Tuple[int, ...]`. However, as this is slightly trickier, it's okay to add a comment here about this and create a follow-up issue. Which would you prefer?

Implementing the general case shouldn't be hard, though it's not urgent really. The meet of arbitrary `Tuple[t1, <...>,  tn]` and `Tuple[s, ...]` (where `<...>` would not be a Python ellipsis but a mathematical one) would be `Tuple[meet(t1, s), ..., meet(tn, s)]`, I think. So the meet of types `Tuple[X, Y]` and `Tuple[Z, ...]` would be `Tuple[meet(X, Z), meet(Y, Z)]`, etc.

JukkaL commented on test-data/unit/check-tuples.test:
```diff
@@ -808,6 +808,73 @@ b = (1, 'x')
+
+[builtins fixtures/tuple.pyi]
+[out]
+main: note: In function "f":
```
You can use `# flags: --hide-error-context` to hide the function names. Then you can use `# E: ...` for multiple functions in a single test case.

ilevkivskyi replied:
@JukkaL Thanks for useful tip!

JukkaL commented on test-data/unit/check-tuples.test:
```diff
@@ -808,6 +808,73 @@ b = (1, 'x')
+        reveal_type(v) # E: Revealed type is 'builtins.int'
+[builtins fixtures/tuple.pyi]
+[out]
+main: note: In function "tuplify":
```
Again this would be clearer if you'd use `# flags: --hide-error-context`. It should perhaps be on by default for test cases, but we have a huge number of test cases that would be affected so it's unclear whether it would be worth the code churn.

JukkaL commented:
Looks good to merge, other than the above issues.

ilevkivskyi commented:
@JukkaL Thank you for review! I implemented your comments and added another small test that triggers the code for ``meet(Tuple[t1, t2], Tuple[s, ...])``.

JukkaL commented:
Looks good now. Thanks!
